src/query/api/v1/handler/options.go

```diff
@@ -0,0 +1,135 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package handler
+
+import (
+	"errors"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/m3db/m3/src/cluster/services"
+)
+
+const (
+	// M3DBServiceName is the service name for M3DB.
+	M3DBServiceName = "m3db"
+	// M3AggregatorServiceName is the service name for M3Aggregator.
+	M3AggregatorServiceName = "m3aggregator"
+	// M3CoordinatorServiceName is the service name for M3Coordinator.
+	M3CoordinatorServiceName = "m3coordinator"
+
+	defaultM3AggMaxAggregationWindowSize = time.Minute
+	// defaultM3AggWarmupDuration configures the buffer to account for the delay
+	// of propagating aggregator placement to clients, usually needed when there is
+	// a large amount of clients sending traffic to m3aggregator.
+	defaultM3AggWarmupDuration = 0
+)
+
+var (
+	errServiceNameIsRequired        = errors.New("service name is required")
+	errServiceEnvironmentIsRequired = errors.New("service environment is required")
+	errServiceZoneIsRequired        = errors.New("service zone is required")
+	errM3AggServiceOptionsRequired  = errors.New("m3agg service options are required")
+)
+
+// ServiceOptions are the options for Service.
+type ServiceOptions struct {
+	ServiceName        string
+	ServiceEnvironment string
+	ServiceZone        string
+
+	M3Agg *M3AggServiceOptions
+
+	DryRun bool
+}
+
+// M3AggServiceOptions contains the service options that are
+// specific to the M3Agg service.
+type M3AggServiceOptions struct {
+	MaxAggregationWindowSize time.Duration
+	WarmupDuration           time.Duration
+}
+
+// NewServiceOptions returns a ServiceOptions based on the provided
+// values.
+func NewServiceOptions(
+	serviceName string, headers http.Header, m3AggOpts *M3AggServiceOptions) ServiceOptions {
+	opts := ServiceOptions{
+		ServiceName:        serviceName,
+		ServiceEnvironment: DefaultServiceEnvironment,
+		ServiceZone:        DefaultServiceZone,
+
+		DryRun: false,
+
+		M3Agg: &M3AggServiceOptions{
+			MaxAggregationWindowSize: defaultM3AggMaxAggregationWindowSize,
+			WarmupDuration:           defaultM3AggWarmupDuration,
+		},
+	}
+
+	if v := strings.TrimSpace(headers.Get(HeaderClusterEnvironmentName)); v != "" {
+		opts.ServiceEnvironment = v
+	}
+	if v := strings.TrimSpace(headers.Get(HeaderClusterZoneName)); v != "" {
+		opts.ServiceZone = v
+	}
+	if v := strings.TrimSpace(headers.Get(HeaderDryRun)); v == "true" {
+		opts.DryRun = true
+	}
+
+	if m3AggOpts != nil {
+		if m3AggOpts.MaxAggregationWindowSize > 0 {
+			opts.M3Agg.MaxAggregationWindowSize = m3AggOpts.MaxAggregationWindowSize
+		}
+
+		if m3AggOpts.WarmupDuration > 0 {
+			opts.M3Agg.WarmupDuration = m3AggOpts.WarmupDuration
+		}
+	}
+
+	return opts
+}
+
+// Validate ensures the service options are valid.
+func (opts *ServiceOptions) Validate() error {
+	if opts.ServiceName == "" {
```

Comment:
Can we have a slice of allowed service names and do that validation here

Reply:
I had trouble with this as well. We have the concept of "allowed services" in the placement handler: https://github.com/m3db/m3/blob/f8508a401e1055c7d7fdaee22982fd48858b3e37/src/query/api/v1/handler/placement/common.go#L151

I tried moving this to the common handler package, but it's used in a few other places in the placement handler. Since it turns out that the placement handler is the only thing that cares about service name (see above comment), I left it there since namespace doesn't need it and I didn't want to have to expose it. Can change that depending on how we feel about having to expose it or not

Reply:
I would just add a new slice in this package called `ValidServices` or something and put every M3Coordinator/M3Aggregator/M3DB in it and call it a day. It would just be nice to know that anywhere we're using this thing that value won't be garbage